From 15fb0e84e38c2681e855e69b58414ba831b399bf Mon Sep 17 00:00:00 2001
From: "Dr. David Alan Gilbert" <dgilbert@redhat.com>
Date: Mon, 9 Jul 2018 19:57:16 +0100
Subject: [PATCH 11/29] DAX: virtiofsd: Wire up passthrough_ll's
 lo_setupmapping

Wire up passthrough_ll's setupmapping to allocate, send to virtio
and then reply OK.

Signed-off-by: Dr. David Alan Gilbert <dgilbert@redhat.com>
Signed-off-by: Vivek Goyal <vgoyal@redhat.com>
---
 tools/virtiofsd/passthrough_ll.c | 24 ++++++++++++++++++++++--
 1 file changed, 22 insertions(+), 2 deletions(-)

diff --git a/tools/virtiofsd/passthrough_ll.c b/tools/virtiofsd/passthrough_ll.c
index 784bdcff34..b57cb4079e 100644
--- a/tools/virtiofsd/passthrough_ll.c
+++ b/tools/virtiofsd/passthrough_ll.c
@@ -2895,8 +2895,28 @@ static void lo_setupmapping(fuse_req_t req, fuse_ino_t ino, uint64_t foffset,
                             uint64_t len, uint64_t moffset, uint64_t flags,
                             struct fuse_file_info *fi)
 {
-    // TODO
-    fuse_reply_err(req, ENOSYS);
+    int ret = 0;
+    VhostUserFSSlaveMsg msg = { 0 };
+    uint64_t vhu_flags;
+    bool writable = flags & O_RDWR;
+
+    vhu_flags = VHOST_USER_FS_FLAG_MAP_R;
+    if (writable) {
+        vhu_flags |= VHOST_USER_FS_FLAG_MAP_W;
+    }
+
+    msg.fd_offset[0] = foffset;
+    msg.len[0] = len;
+    msg.c_offset[0] = moffset;
+    msg.flags[0] = vhu_flags;
+
+    if (fuse_virtio_map(req, &msg, lo_fi_fd(req, fi))) {
+        fprintf(stderr, "%s: map over virtio failed (fd=%d)\n", __func__,
+                (int)fi->fh);
+        ret = EINVAL;
+    }
+
+    fuse_reply_err(req, ret);
 }
 
 static void lo_removemapping(fuse_req_t req, struct fuse_session *se,
-- 
2.25.1

